server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java

```diff
@@ -0,0 +1,348 @@
+        if (timedOut == false) {
+            for (final PublicationTarget target : publicationTargets) {
+                if (target.isActive()) {
+                    return isCompleted == false;
+                }
+            }
+        }
+        return isCompleted;
+    }
+
+    private void onPossibleCommitFailure() {
+        if (applyCommitRequest.isPresent()) {
+            onPossibleCompletion();
+            return;
+        }
+
+        final CoordinationState.VoteCollection possiblySuccessfulNodes = new CoordinationState.VoteCollection();
+        for (PublicationTarget publicationTarget : publicationTargets) {
+            if (publicationTarget.mayCommitInFuture()) {
+                possiblySuccessfulNodes.addVote(publicationTarget.discoveryNode);
+            } else {
+                assert publicationTarget.isFailed() : publicationTarget;
+            }
+        }
+
+        if (isPublishQuorum(possiblySuccessfulNodes) == false) {
+            logger.debug("onPossibleCommitFailure: non-failed nodes {} do not form a quorum, so {} cannot succeed",
+                possiblySuccessfulNodes, this);
+            Exception e = new Discovery.FailedToCommitClusterStateException("non-failed nodes do not form a quorum");
+            publicationTargets.stream().filter(PublicationTarget::isActive).forEach(pt -> pt.setFailed(e));
+            onPossibleCompletion();
+        }
+    }
+
+    protected abstract void onCompletion(boolean committed);
+
+    protected abstract boolean isPublishQuorum(CoordinationState.VoteCollection votes);
+
+    protected abstract Optional<ApplyCommitRequest> handlePublishResponse(DiscoveryNode sourceNode, PublishResponse publishResponse);
+
+    protected abstract void onPossibleJoin(DiscoveryNode sourceNode, PublishWithJoinResponse response);
+
+    protected abstract void sendPublishRequest(DiscoveryNode destination, PublishRequest publishRequest,
+                                               ActionListener<PublishWithJoinResponse> responseActionListener);
+
+    protected abstract void sendApplyCommit(DiscoveryNode destination, ApplyCommitRequest applyCommit,
+                                            ActionListener<TransportResponse.Empty> responseActionListener);
+
+    @Override
+    public String toString() {
+        return "Publication{term=" + publishRequest.getAcceptedState().term() +
+            ", version=" + publishRequest.getAcceptedState().version() + '}';
+    }
+
+    enum PublicationTargetState {
+        NOT_STARTED,
+        FAILED,
+        SENT_PUBLISH_REQUEST,
+        WAITING_FOR_QUORUM,
+        SENT_APPLY_COMMIT,
+        APPLIED_COMMIT,
+    }
+
+    private class PublicationTarget {
+        private final DiscoveryNode discoveryNode;
+        private boolean ackIsPending = true;
+        private PublicationTargetState state = PublicationTargetState.NOT_STARTED;
+
+        private PublicationTarget(DiscoveryNode discoveryNode) {
```

Comment:
Not 100% sure of the rules here, but `private` seems too restricted for this.

Reply:
fixed by 799157d72d4